sql: miscellaneous improvements to reduce allocations

yuzefovich opened the pull request:
See individual commits for details.

All 6 commits together provide the following improvement
```
name                                           old time/op    new time/op    delta
FlowSetup/vectorize=true/distribute=true-24       168µs ± 5%     162µs ± 5%  -3.56%  (p=0.000 n=20+20)
FlowSetup/vectorize=true/distribute=false-24      162µs ± 5%     159µs ± 3%  -1.89%  (p=0.028 n=17+19)
FlowSetup/vectorize=false/distribute=true-24      160µs ± 2%     162µs ± 5%    ~     (p=0.104 n=17+20)
FlowSetup/vectorize=false/distribute=false-24     156µs ± 4%     158µs ± 5%    ~     (p=0.057 n=19+19)

name                                           old alloc/op   new alloc/op   delta
FlowSetup/vectorize=true/distribute=true-24      18.9kB ± 2%    18.4kB ± 5%  -2.78%  (p=0.000 n=17+17)
FlowSetup/vectorize=true/distribute=false-24     17.7kB ± 2%    17.2kB ± 2%  -3.11%  (p=0.000 n=16+17)
FlowSetup/vectorize=false/distribute=true-24     25.4kB ± 1%    25.3kB ± 0%  -0.58%  (p=0.000 n=17+17)
FlowSetup/vectorize=false/distribute=false-24    24.3kB ± 0%    24.5kB ± 6%  +0.51%  (p=0.019 n=17+18)

name                                           old allocs/op  new allocs/op  delta
FlowSetup/vectorize=true/distribute=true-24         206 ± 2%       192 ± 2%  -6.35%  (p=0.000 n=20+20)
FlowSetup/vectorize=true/distribute=false-24        196 ± 2%       182 ± 2%  -7.17%  (p=0.000 n=18+19)
FlowSetup/vectorize=false/distribute=true-24        197 ± 0%       193 ± 0%  -2.03%  (p=0.000 n=16+17)
FlowSetup/vectorize=false/distribute=false-24       188 ± 0%       183 ± 1%  -2.63%  (p=0.000 n=17+16)
```

Release justification: low-risk improvement.

Release note: None

yuzefovich committed:
sql: remove an allocation in checkScanParallelizationIfLocal

```
name                                           old time/op    new time/op    delta
FlowSetup/vectorize=true/distribute=true-24       167µs ± 4%     165µs ± 4%    ~     (p=0.327 n=20+20)
FlowSetup/vectorize=true/distribute=false-24      160µs ± 4%     164µs ±11%    ~     (p=0.515 n=18+20)
FlowSetup/vectorize=false/distribute=true-24      160µs ± 6%     159µs ± 4%    ~     (p=0.298 n=19+18)
FlowSetup/vectorize=false/distribute=false-24     158µs ± 7%     157µs ± 7%    ~     (p=0.841 n=20+20)

name                                           old alloc/op   new alloc/op   delta
FlowSetup/vectorize=true/distribute=true-24      18.9kB ± 2%    19.0kB ± 6%    ~     (p=0.883 n=17+18)
FlowSetup/vectorize=true/distribute=false-24     17.7kB ± 1%    17.7kB ± 1%    ~     (p=0.978 n=16+16)
FlowSetup/vectorize=false/distribute=true-24     25.5kB ± 3%    25.4kB ± 2%    ~     (p=0.702 n=18+16)
FlowSetup/vectorize=false/distribute=false-24    24.4kB ± 0%    24.4kB ± 2%    ~     (p=0.545 n=16+16)

name                                           old allocs/op  new allocs/op  delta
FlowSetup/vectorize=true/distribute=true-24         205 ± 1%       205 ± 2%    ~     (p=0.464 n=20+19)
FlowSetup/vectorize=true/distribute=false-24        197 ± 3%       196 ± 1%  -0.62%  (p=0.013 n=19+19)
FlowSetup/vectorize=false/distribute=true-24        197 ± 0%       197 ± 0%    ~     (p=1.000 n=17+17)
FlowSetup/vectorize=false/distribute=false-24       188 ± 0%       187 ± 1%  -0.50%  (p=0.000 n=16+16)
```

Release justification: low-risk improvement.

Release note: None

yuzefovich committed:
colflow: remove recently introduced allocation

This commit removes recently introduced allocation for an anonymous
factory method to create multiple streaming accounts. We can just pass
in the monitor registry directly and use it only if needed.

Additionally it fixes an incorrect error propagation in the flow setup
(which wasn't a big deal since an error could only occur if an assertion
is failed).

```
name                                           old time/op    new time/op    delta
FlowSetup/vectorize=true/distribute=true-24       162µs ± 3%     163µs ± 6%    ~     (p=0.629 n=16+19)
FlowSetup/vectorize=true/distribute=false-24      160µs ± 6%     163µs ±11%    ~     (p=0.740 n=18+20)
FlowSetup/vectorize=false/distribute=true-24      164µs ± 8%     160µs ± 6%  -2.63%  (p=0.005 n=19+19)
FlowSetup/vectorize=false/distribute=false-24     158µs ± 5%     157µs ±11%    ~     (p=0.065 n=19+20)

name                                           old alloc/op   new alloc/op   delta
FlowSetup/vectorize=true/distribute=true-24      18.9kB ± 2%    19.0kB ± 7%    ~     (p=0.935 n=16+19)
FlowSetup/vectorize=true/distribute=false-24     17.7kB ± 1%    17.7kB ± 1%    ~     (p=0.127 n=17+16)
FlowSetup/vectorize=false/distribute=true-24     25.4kB ± 0%    25.4kB ± 0%    ~     (p=0.936 n=16+17)
FlowSetup/vectorize=false/distribute=false-24    24.4kB ± 1%    24.4kB ± 0%    ~     (p=0.713 n=19+18)

name                                           old allocs/op  new allocs/op  delta
FlowSetup/vectorize=true/distribute=true-24         205 ± 1%       203 ± 2%  -1.01%  (p=0.000 n=19+20)
FlowSetup/vectorize=true/distribute=false-24        196 ± 1%       194 ± 2%  -0.92%  (p=0.013 n=19+19)
FlowSetup/vectorize=false/distribute=true-24        197 ± 0%       197 ± 0%    ~     (all equal)
FlowSetup/vectorize=false/distribute=false-24       187 ± 0%       187 ± 0%    ~     (all equal)
```

Release justification: low-risk cleanup.

Release note: None

yuzefovich committed:
colbuilder: reuse the unlimited monitors for the same operators

This commit reduces the number of monitors that we create for different
operators. Previously, for each unlimited memory account we would
instantiate a separate memory monitor, but that doesn't give us much
benefit and incurs extra allocations, so this commit reuses the same
unlimited monitor for as many accounts as needed for a component.

Additionally this commit reduces the allocations when constructing
a unique memory monitor name which were introduced when we marked the
monitor name as `redact.RedactableString`.

```
name                                           old time/op    new time/op    delta
FlowSetup/vectorize=true/distribute=true-24       162µs ± 4%     160µs ± 6%    ~     (p=0.087 n=18+20)
FlowSetup/vectorize=true/distribute=false-24      162µs ± 4%     161µs ± 5%    ~     (p=0.181 n=19+19)
FlowSetup/vectorize=false/distribute=true-24      159µs ± 4%     160µs ± 7%    ~     (p=0.862 n=20+20)
FlowSetup/vectorize=false/distribute=false-24     160µs ±11%     155µs ± 6%  -3.40%  (p=0.009 n=20+20)

name                                           old alloc/op   new alloc/op   delta
FlowSetup/vectorize=true/distribute=true-24      18.9kB ± 5%    18.8kB ± 6%    ~     (p=0.097 n=18+20)
FlowSetup/vectorize=true/distribute=false-24     17.7kB ± 1%    17.5kB ± 7%  -0.82%  (p=0.001 n=17+18)
FlowSetup/vectorize=false/distribute=true-24     25.4kB ± 0%    25.4kB ± 0%    ~     (p=0.824 n=16+17)
FlowSetup/vectorize=false/distribute=false-24    24.4kB ± 0%    24.4kB ± 1%    ~     (p=0.558 n=16+16)

name                                           old allocs/op  new allocs/op  delta
FlowSetup/vectorize=true/distribute=true-24         203 ± 1%       196 ± 2%  -3.39%  (p=0.000 n=19+19)
FlowSetup/vectorize=true/distribute=false-24        194 ± 1%       187 ± 3%  -3.59%  (p=0.000 n=19+19)
FlowSetup/vectorize=false/distribute=true-24        197 ± 0%       197 ± 0%    ~     (all equal)
FlowSetup/vectorize=false/distribute=false-24       187 ± 0%       187 ± 0%    ~     (p=0.515 n=16+17)
```

Release justification: low-risk improvement.

Release note: None

yuzefovich committed:
sql: remove DatumAlloc from planner

This commit removes the shared `tree.DatumAlloc` from the `planner`
object since it is barely used nowadays. Instead, it is constructed
on-demand in the exec factory.

Release justification: low-risk cleanup.

Release note: None

yuzefovich committed:
sql: remove some allocations around partition handling

This commit removes some allocations around handling of the partitions
(driven by the heap profile of `newOptTable`):
- it introduces a couple of methods to `catalog.Index` about the number
of columns and the number of implicit columns in the partitioning. The
idea is to avoid an allocation of the `partitioning` struct when only an
integer result is desired
- it lazily allocates and reuses a map in `newOptTable` around zones of
the partitions.

```
name                                           old time/op    new time/op    delta
FlowSetup/vectorize=true/distribute=true-24       158µs ± 5%     159µs ± 4%    ~     (p=0.538 n=19+18)
FlowSetup/vectorize=true/distribute=false-24      157µs ± 8%     156µs ± 5%    ~     (p=0.563 n=19+19)
FlowSetup/vectorize=false/distribute=true-24      160µs ± 5%     161µs ± 6%    ~     (p=0.565 n=20+20)
FlowSetup/vectorize=false/distribute=false-24     157µs ± 7%     157µs ± 8%    ~     (p=0.922 n=19+18)

name                                           old alloc/op   new alloc/op   delta
FlowSetup/vectorize=true/distribute=true-24      18.8kB ± 8%    18.7kB ± 7%    ~     (p=0.277 n=20+20)
FlowSetup/vectorize=true/distribute=false-24     17.3kB ± 1%    17.2kB ± 1%    ~     (p=0.052 n=17+18)
FlowSetup/vectorize=false/distribute=true-24     25.4kB ± 0%    25.3kB ± 0%  -0.29%  (p=0.000 n=16+16)
FlowSetup/vectorize=false/distribute=false-24    24.4kB ± 1%    24.6kB ± 6%    ~     (p=0.905 n=16+18)

name                                           old allocs/op  new allocs/op  delta
FlowSetup/vectorize=true/distribute=true-24         196 ± 2%       194 ± 2%  -1.34%  (p=0.000 n=19+18)
FlowSetup/vectorize=true/distribute=false-24        186 ± 1%       185 ± 3%  -0.89%  (p=0.022 n=19+20)
FlowSetup/vectorize=false/distribute=true-24        197 ± 0%       195 ± 0%  -1.08%  (p=0.000 n=16+16)
FlowSetup/vectorize=false/distribute=false-24       187 ± 0%       185 ± 0%  -1.07%  (p=0.000 n=16+16)
```

Release justification: low-risk improvement.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86962)
<!-- Reviewable:end -->

yuzefovich committed:
util/mon: remove an allocation for "no reserved" account

This commit introduces a singleton "no reserved" account when a monitor
is started with no reservation up-front. This account can be shared with
all such monitors since they only will ever read the "used" field.

```
name                                           old time/op    new time/op    delta
FlowSetup/vectorize=true/distribute=true-24       165µs ± 5%     163µs ± 5%    ~     (p=0.066 n=19+18)
FlowSetup/vectorize=true/distribute=false-24      165µs ± 2%     163µs ± 9%    ~     (p=0.097 n=17+20)
FlowSetup/vectorize=false/distribute=true-24      164µs ± 3%     164µs ± 5%    ~     (p=0.425 n=18+19)
FlowSetup/vectorize=false/distribute=false-24     164µs ±12%     160µs ± 5%    ~     (p=0.277 n=20+20)

name                                           old alloc/op   new alloc/op   delta
FlowSetup/vectorize=true/distribute=true-24      18.6kB ± 7%    18.3kB ± 3%  -1.69%  (p=0.014 n=19+17)
FlowSetup/vectorize=true/distribute=false-24     17.3kB ± 1%    17.3kB ± 8%  -0.01%  (p=0.007 n=17+17)
FlowSetup/vectorize=false/distribute=true-24     25.5kB ± 3%    25.3kB ± 0%  -0.77%  (p=0.000 n=17+17)
FlowSetup/vectorize=false/distribute=false-24    24.3kB ± 1%    24.2kB ± 1%  -0.45%  (p=0.001 n=16+16)

name                                           old allocs/op  new allocs/op  delta
FlowSetup/vectorize=true/distribute=true-24         195 ± 3%       191 ± 0%  -1.62%  (p=0.000 n=19+16)
FlowSetup/vectorize=true/distribute=false-24        185 ± 2%       183 ± 3%  -1.02%  (p=0.010 n=18+19)
FlowSetup/vectorize=false/distribute=true-24        195 ± 0%       193 ± 0%  -1.06%  (p=0.000 n=16+17)
FlowSetup/vectorize=false/distribute=false-24       185 ± 0%       183 ± 0%  -1.08%  (p=0.000 n=16+16)
```

Release justification: low-risk cleanup.

Release note: None

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r1, 3 of 3 files at r2, 3 of 3 files at r3, 7 of 7 files at r4, 12 of 12 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86962#-:-NASNpEc9t_00VS5bTJx:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @msirek and @yuzefovich)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86962#-NASIhxDEHbwm8aiEv2J):*
Is the difference here because previously the two bool variables had to be allocated on the heap, and now they both can be allocated at once within the struct?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86962#-:-NAVGSv-Elkq67jw3lsN:b-oduykh)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @msirek)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86962#-NASIhxDEHbwm8aiEv2J):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Is the difference here because previously the two bool variables had to be allocated on the heap, and now they both can be allocated at once within the struct?
</blockquote></details>

I think the difference is that we previously had to allocate an anonymous function on the heap and now we don't. It's quite possible that those two return bool variables prohibited the compiler from proving that the function didn't outlive `checkScanParallelizationIfLocal`, so it didn't have to be put onto the heap.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
bors r+
